File: Data_structures/Using_C/Ch-01/BSWorstOpCount.c
#include <stdio.h>

int
BSearch (int  ar[],
         int  len,
         int  target)
{
  int first   = 0;
  int last    = len - 1;
  int opCount = 0;
  int mid;

  while (first <= last)
    {
      mid = (first + last) / 2;

      if (target == ar[mid])
        {
          return mid;
        }
      else
        {
          if (target < ar[mid])
            last = mid - 1;
          else
            first = mid + 1;
        }

      opCount += 1;
    }
  printf ("비교연산횟수: %d \n", opCount);

  return -1;
}


int
main (int   argc,
      char *argv[])
{
  int arr1[500]   = {0,};
  int arr2[5000]  = {0,};
  int arr3[50000] = {0,};
  int idx;

  idx = BSearch (arr1, sizeof (arr1) / sizeof (int), 1);
  if (idx == -1)
    printf ("탐색 실패 \n\n");
  else
    printf ("타겟 저장 인덱스: %d \n", idx);

  idx = BSearch (arr2, sizeof (arr2) / sizeof (int), 2);
  if (idx == -1)
    printf ("탐색 실패 \n\n");
  else
    printf ("타겟 저장 인덱스: %d \n", idx);

  idx = BSearch (arr3, sizeof (arr3) / sizeof (int), 3);
  if (idx == -1)
    printf ("탐색 실패 \n\n");
  else
    printf ("타겟 저장 인덱스: %d \n", idx);

  return 0;
}

File: Data_structures/Using_perl/Ch-01/BSWorstOpCount.pl
#!/usr/bin/env perl

use 5.014;
use strict;
use warnings;

my @arr1 = map {$_ * 0} (1 .. 500);
my @arr2 = map {$_ * 0} (1 .. 5000);
my @arr3 = map {$_ * 0} (1 .. 50000);
my $idx;

$idx = BSearch (\@arr1, scalar @arr1, 1);
$idx == -1 ?  say "탐색실패\n" : say "타겟 저장 인덱스: $idx";

$idx = BSearch (\@arr2, scalar @arr2, 2);
$idx == -1 ?  say "탐색실패\n" : say "타겟 저장 인덱스: $idx";

$idx = BSearch (\@arr3, scalar @arr3, 3);
$idx == -1 ?  say "탐색실패\n" : say "타겟 저장 인덱스: $idx";

sub BSearch {
    my ( $arrs, $len, $target ) = @_;

    my $first = 0;
    my $last = $len - 1;
    my $opCount = 0;
    my $mid;

    while ( $first <= $last ) {
        $mid = int ( ($first + $last) / 2 );

        if ($target == $arrs->[$mid])
        {
            return $mid;
        }
        else {
            $target < $arrs->[$mid] ? ($last = $mid - 1) : ($first = $mid + 1);
        }

        $opCount += 1;
    }
    say "비교 연산 횟수: $opCount";

    return -1;
}
